File: simulation/격자 안에서 단일 객체를 이동/숫자가_더_큰_인접한_곳으로_이동.java
import java.util.*;
import java.io.*;

public class Main {
    static final int MAX_N = 100;
    static final int DIR_NUM = 4;

    static int n;
    static int currX, currY;
    static int[][] grid = new int[MAX_N + 1][MAX_N + 1];

    static List<Integer> visitedNums = new ArrayList<>();

    static boolean canGo(int x, int y, int currNum) {
        if (1 <= x && x <= n && 1 <= y && y <= n) {
            if (grid[x][y] > currNum) {
                return true;
            }
        }
        return false;
    }

    static boolean simulate() {
        int[] dx = new int[]{-1, 1, 0, 0};
        int[] dy = new int[]{0, 0, -1, 1};

        for (int i = 0; i < 4; i++) {
            int nextX = currX + dx[i];
            int nextY = currY + dy[i];

            if (canGo(nextX, nextY, grid[currX][currY])) {
                currX = nextX;
                currY = nextY;
                return true;
            }
        }
        return false;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        currX = Integer.parseInt(st.nextToken());
        currY = Integer.parseInt(st.nextToken());

        for (int i = 1; i <= n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 1; j <= n; j++) {
                grid[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        visitedNums.add(grid[currX][currY]);

        while (true) {
            boolean greaterNumberExist = simulate();

            if (!greaterNumberExist) {
                break;
            }

            visitedNums.add(grid[currX][currY]);
        }

        for (int i = 0; i < visitedNums.size(); i++) {
            System.out.print(visitedNums.get(i) + " ");
        }
    }
}

File: Park/CodeTree/격자 안에서 단일 객체를 이동/숫자가 더 큰 인접한 곳으로 이동.py
# 변수 선언 및 입력
n, curr_x, curr_y = map(int, input().split())
a = [[0] * (n + 1)] + [[0] + list(map(int, input().split())) for _ in range(n)]

# 방문한 숫자를 담을 리스트
visited_nums = []

# 범위 확인 함수
def in_range(x, y):
    return 1 <= x <= n and 1 <= y <= n

# 이동 가능 여부 확인 함수
def can_go(x, y, curr_num):
    return in_range(x, y) and a[x][y] > curr_num

# 시뮬레이션 함수
def simulate():
    global curr_x, curr_y
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    
    for dx, dy in directions:
        next_x, next_y = curr_x + dx, curr_y + dy
        if can_go(next_x, next_y, a[curr_x][curr_y]):
            curr_x, curr_y = next_x, next_y
            return True
    return False

# 초기 위치 값 추가
visited_nums.append(a[curr_x][curr_y])

# 시뮬레이션 실행
while simulate():
    visited_nums.append(a[curr_x][curr_y])

# 결과 출력
print(' '.join(map(str, visited_nums)))
